File: app/src/main/java/ru/art2000/calculator/view/calculator/HistoryListAdapter.java
package ru.art2000.calculator.view.calculator;

import android.content.Context;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import ru.art2000.calculator.R;
import ru.art2000.calculator.model.calculator.HistoryItem;
import ru.art2000.calculator.view_model.calculator.HistoryViewModel;
import ru.art2000.extensions.collections.CollectionsKt;

public class HistoryListAdapter extends RecyclerView.Adapter<HistoryListAdapter.ViewHolder> {

    private final LayoutInflater mInflater;
    private final Context mContext;
    private final HistoryViewModel mModel;

    private List<HistoryItem> historyList = new ArrayList<>();

    HistoryListAdapter(Context context, LifecycleOwner lifecycleOwner, HistoryViewModel model, LiveData<List<HistoryItem>> items) {
        mInflater = LayoutInflater.from(context);
        mContext = context;
        mModel = model;
        items.observe(lifecycleOwner, this::setNewData);
    }

    public List<HistoryItem> getHistoryList() {
        return historyList;
    }

    private void setNewData(@NonNull List<HistoryItem> newData) {
        if (historyList == null || historyList.isEmpty()) {
            historyList = newData;
            notifyItemRangeInserted(0, newData.size());
        } else {
            DiffUtil.DiffResult result = CollectionsKt.calculateDiff(historyList, newData);

            historyList = newData;
            result.dispatchUpdatesTo(this);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.item_history_list, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final ViewHolder holder, int position) {
        HistoryItem item = historyList.get(position);
        holder.expr.setText(item.getExpression());
        holder.res.setText(item.getResult());
    }

    @Override
    public void onViewRecycled(@NonNull ViewHolder holder) {
        holder.itemView.setOnLongClickListener(null);
        super.onViewRecycled(holder);
    }

    @Override
    public int getItemCount() {
        return historyList.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnCreateContextMenuListener {
        TextView expr;
        TextView res;

        ViewHolder(View itemView) {
            super(itemView);
            expr = itemView.findViewById(R.id.expression);
            res = itemView.findViewById(R.id.result);
            itemView.setOnCreateContextMenuListener(this);
        }

        @Override
        public void onCreateContextMenu(ContextMenu menu, View v,
                                        ContextMenu.ContextMenuInfo menuInfo) {

            menu.setHeaderTitle(mContext.getString(R.string.you_can));

            menu.add(Menu.NONE, HistoryViewModel.COPY_ALL, Menu.NONE, mContext.getString(R.string.context_menu_copy_all)).setOnMenuItemClickListener(this::onContextItemSelected);
            menu.add(Menu.NONE, HistoryViewModel.COPY_EXPR, Menu.NONE, mContext.getString(R.string.context_menu_copy_expr)).setOnMenuItemClickListener(this::onContextItemSelected);
            menu.add(Menu.NONE, HistoryViewModel.COPY_RES, Menu.NONE, mContext.getString(R.string.context_menu_copy_res)).setOnMenuItemClickListener(this::onContextItemSelected);
            menu.add(Menu.NONE, HistoryViewModel.DELETE, Menu.NONE, mContext.getString(R.string.delete_record)).setOnMenuItemClickListener(this::onContextItemSelected);
            menu.removeItem(HistoryViewModel.DELETE);
        }

        private boolean onContextItemSelected(@NonNull MenuItem menuItem) {

            int id = menuItem.getItemId();

            String toastText;
            if (id >= HistoryViewModel.COPY_ALL && id <= HistoryViewModel.COPY_RES) {

                HistoryItem selectedItem = historyList.get(getBindingAdapterPosition());
                toastText = mModel.copyHistoryItemToClipboard(selectedItem, id);

            } else if (id == HistoryViewModel.DELETE) {

                HistoryItem selectedItem = historyList.get(getBindingAdapterPosition());
                mModel.removeHistoryItem(selectedItem.getId());
                toastText = mContext.getString(R.string.deleted) + " " + selectedItem.getFullExpression();

            } else return true;

            Toast.makeText(mContext, toastText, Toast.LENGTH_SHORT).show();

            return true;
        }

    }
}

File: app/src/main/java/ru/art2000/calculator/view/calculator/HistoryListAdapter.kt
package ru.art2000.calculator.view.calculator

import android.content.Context
import android.os.Build
import android.view.*
import android.view.ContextMenu.ContextMenuInfo
import android.view.View.OnCreateContextMenuListener
import android.widget.EditText
import android.widget.Toast
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.flow.Flow
import ru.art2000.calculator.R
import ru.art2000.calculator.databinding.ItemHistoryDateBinding
import ru.art2000.calculator.databinding.ItemHistoryListBinding
import ru.art2000.calculator.model.calculator.history.HistoryDatabaseItem
import ru.art2000.calculator.model.calculator.history.HistoryDateItem
import ru.art2000.calculator.model.calculator.history.HistoryListItem
import ru.art2000.calculator.model.calculator.history.HistoryValueItem
import ru.art2000.calculator.view.calculator.HistoryListAdapter.HistoryViewHolder
import ru.art2000.calculator.view_model.calculator.HistoryViewModel
import ru.art2000.extensions.arch.launchAndCollect
import ru.art2000.extensions.arch.launchRepeatOnStarted
import ru.art2000.extensions.collections.calculateDiff
import ru.art2000.extensions.views.textValue
import ru.art2000.extensions.views.toViewString

class HistoryListAdapter internal constructor(
        private val context: Context,
        lifecycleOwner: LifecycleOwner,
        private val model: HistoryViewModel,
        items: Flow<List<HistoryListItem>>
) : RecyclerView.Adapter<HistoryViewHolder>() {

    var historyList: List<HistoryListItem> = listOf()
        private set

    private val inflater = LayoutInflater.from(context)

    private fun setNewData(newData: List<HistoryListItem>) {
        if (historyList.isEmpty()) {
            historyList = newData
            notifyItemRangeInserted(0, newData.size)
        } else {
            val result = calculateDiff(historyList, newData)
            historyList = newData
            result.dispatchUpdatesTo(this)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return when(historyList[position]) {
            is HistoryValueItem -> VALUE_ITEM
            is HistoryDateItem -> DATE_ITEM
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = when (viewType) {
        VALUE_ITEM -> {
            val binding = ItemHistoryListBinding.inflate(inflater, parent, false)
            ValueViewHolder(binding)
        }
        DATE_ITEM -> {
            val binding = ItemHistoryDateBinding.inflate(inflater, parent, false)
            DateViewHolder(binding)
        }
        else -> throw IllegalStateException("Unknown view type '$viewType'")
    }

    override fun onBindViewHolder(holder: HistoryViewHolder, position: Int) {
        val item = historyList[position]
        holder.bind(item)
    }

    override fun onViewRecycled(holder: HistoryViewHolder) {
        holder.itemView.setOnLongClickListener(null)
        super.onViewRecycled(holder)
    }

    override fun getItemCount() = historyList.size

    fun isDateItem(position: Int): Boolean = historyList.getOrNull(position) is HistoryDateItem

    abstract inner class HistoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        abstract fun bind(historyListItem: HistoryListItem)
    }

    inner class ValueViewHolder internal constructor(private val binding: ItemHistoryListBinding) : HistoryViewHolder(binding.root), OnCreateContextMenuListener {

        private val currentDbItem: HistoryDatabaseItem
            get() = (historyList[bindingAdapterPosition] as HistoryValueItem).dbItem

        override fun bind(historyListItem: HistoryListItem) {
            require(historyListItem is HistoryValueItem)

            historyListItem.dbItem.apply {
                binding.expression.text = expression
                binding.result.text = result
                binding.comment.text = comment
                binding.comment.visibility = if (comment.isNullOrEmpty()) View.GONE else View.VISIBLE
            }
        }

        override fun onCreateContextMenu(menu: ContextMenu, v: View,
                                         menuInfo: ContextMenuInfo?) {
            menu.setHeaderTitle(R.string.you_can)

            fun Menu.addHistoryMenuItem(itemId: Int, titleRes: Int) = add(
                    Menu.NONE,
                    itemId,
                    Menu.NONE,
                    titleRes,
            ).setOnMenuItemClickListener {
                onContextItemSelected(it)
            }

            menu.addHistoryMenuItem(HistoryViewModel.COMMENT, R.string.comment_item)
            menu.addHistoryMenuItem(HistoryViewModel.COPY_ALL, R.string.context_menu_copy_all)
            menu.addHistoryMenuItem(HistoryViewModel.COPY_EXPR, R.string.context_menu_copy_expr)
            menu.addHistoryMenuItem(HistoryViewModel.COPY_RES, R.string.context_menu_copy_res)
            menu.addHistoryMenuItem(HistoryViewModel.DELETE, R.string.delete_record)
        }

        private fun onContextItemSelected(menuItem: MenuItem): Boolean {
            val id = menuItem.itemId
            val isCopy = id >= HistoryViewModel.COPY_ALL && id <= HistoryViewModel.COPY_RES
            val toastText = if (isCopy) {
                model.copyHistoryItemToClipboard(currentDbItem, id)
            } else if (id == HistoryViewModel.DELETE) {
                val selectedItem = currentDbItem
                model.removeHistoryItem(selectedItem)
                context.getString(R.string.deleted) + " " + selectedItem.fullExpression
            } else if (id == HistoryViewModel.COMMENT) {
                showCommentDialog()
                null
            } else return true

            // API 33+ features UI showing copied content, so skip toasts for them
            if (!isCopy || Build.VERSION.SDK_INT <= Build.VERSION_CODES.S_V2) {
                toastText?.also { Toast.makeText(context, it, Toast.LENGTH_SHORT).show() }
            }
            return true
        }

        private fun showCommentDialog() {

            val dbItem = currentDbItem
            val editText = EditText(context)
            editText.textValue = dbItem.comment ?: ""

            MaterialAlertDialogBuilder(context)
                .setTitle(R.string.comment_item_title)
                .setView(editText)
                .setNegativeButton(R.string.cancel) { _, _ -> }
                .setPositiveButton(R.string.ok) { _, _ ->
                    dbItem.comment = editText.text.toString().takeIf { it.isNotEmpty() }
                    model.updateHistoryItem(dbItem)
                }.show()
        }

        init {
            itemView.setOnCreateContextMenuListener(this)
        }
    }

    inner class DateViewHolder internal constructor(private val binding: ItemHistoryDateBinding) : HistoryViewHolder(binding.root) {

        override fun bind(historyListItem: HistoryListItem) {
            require(historyListItem is HistoryDateItem)

            binding.date.text = historyListItem.date.toViewString()
        }
    }

    init {
        lifecycleOwner.launchRepeatOnStarted {
            launchAndCollect(items) { setNewData(it) }
        }
    }

    companion object {
        const val VALUE_ITEM = 0
        const val DATE_ITEM = 1
    }
}